types/react-router/index.d.ts

```diff
@@ -128,25 +134,31 @@ export interface match<Params extends { [K in keyof Params]?: string } = {}> {
 // Omit taken from https://github.com/Microsoft/TypeScript/issues/28339#issuecomment-467220238
 export type Omit<T, K extends keyof T> = T extends any ? Pick<T, Exclude<keyof T, K>> : never;
 
+// Newer Omit type: as the previous one is being exported, removing it would be a breaking change
+export type OmitNative<T, K extends string | number | symbol> = { [P in Exclude<keyof T, K>]: T[P]; };
+
 export function matchPath<Params extends { [K in keyof Params]?: string }>(
     pathname: string,
     props: string | string[] | RouteProps,
     parent?: match<Params> | null,
 ): match<Params> | null;
 
-export type ExtractRouteOptionalParam<T extends string> = T extends `${infer Param}?`
-    ? { [k in Param]?: string | number | boolean }
-    : { [k in T]: string | number | boolean };
+export type ExtractRouteOptionalParam<T extends string, U> = T extends `${infer Param}?`
+    ? { [k in Param]?: U }
+    : { [k in T]: U };
```

Comment:
```suggestion
export type ExtractRouteOptionalParam<T extends string, U = string | number | boolean> = T extends `${infer Param}?`
    ? { [k in Param]?: U }
    : { [k in T]: U };
```
otherwise `U` might default to `unknown` if TypeScript can't infer it. And then private and public usage does not need to be changed.

Reply:
thanks! changed related types to follow this 😄